### PR TITLE
ug: op: Update doc with details on rollback

### PR DESCRIPTION
**Pull request by mike-sul**
## Readiness

* [ ] Merge (pending reviews)
* [ ] Merge after _date or event_
* [x] Draft

## Overview

_Why merge this PR? What does it solve?_

## Checklist

_Optional. Add a 'x' to steps taken._
_You can fill this out after opening the PR. "Did I..."_

* [x] Run spelling and grammar check, preferably with linter.
* [x] Avoid changing any header associated with a link/reference.
* [x] Step through instructions (or ask someone to do so).
* [x] Review for [wordiness](https://languagetool.org/insights/post/wordiness/)
* [x] Match tone and style of page/section.
* [x] Run `make linkcheck`.
* [ ] View HTML in a browser to check rendering.
* [ ] Use [semantic newlines](https://bobheadxi.dev/semantic-line-breaks/).
* [ ] follow best practices for commits.
  * [ ] Descriptive title written in the imperative.
  * [ ] Include brief overview of QA steps taken.
  * [ ] Mention any related issues numbers.
  * [ ] End message with sign off/DCO line (`-s, --signoff`).
  * [ ] Sign commit with your gpg key (`-S, --gpg-sign`).
  * [ ] Squash commits if needed.
* [ ] Request PR review by a technical writer and at least one peer.

## Comments

Update the doc on the offline update with details related to boot firmware update and rollback.

### CODE REVIEW

**Comment by doanac** [comment]
Docs for 5c9a9b1da058392603fbe30733f5822b2f9c0f7a are browsable at: https://ci.foundries.io/projects/fio-docs/builds/2003/docs/artifacts/html/index.html

**Comment by doanac** [comment]
Docs for 43306ea8d18560f25c573f17f2fbe1e5a16e5d7a are browsable at: https://ci.foundries.io/projects/fio-docs/builds/2004/docs/artifacts/html/index.html

**Comment by doanac** [comment]
Docs for 3a7755960a3be52e62e1b49e59b5947412aa9c98 are browsable at: https://ci.foundries.io/projects/fio-docs/builds/2005/docs/artifacts/html/index.html

**Comment by mike-sul** [comment]
I am not sure if we should mentioned the ``110`` and ``120`` codes at all, since it should not happen under normal circumstances and should be considered as a general unknown error.

Another option to recover from rollback to "unknown" is to carry/have two sets of source content, one with the current version (but fresh TUF meta) and the second with the desired version.  So, if things go really bad during an update to the desired version a user can always update to the previous/current version manually. I just don't know if we should mention it in the doc neither.

File: source/user-guide/offline-update/offline-update.rst [hunk 1]
```diff
@@ -55,9 +55,16 @@ Use the ``aklite-offline`` CLI utility to perform an offline update:If a device
 
 2. Run one of the post installation actions depending on the ``aklite-offline`` install result:
```
Comment (doanac) [inline source/user-guide/offline-update/offline-update.rst:56]: maybe note the return codes are documented below?

File: source/user-guide/offline-update/offline-update.rst [hunk 2]
```diff
@@ -78,16 +85,47 @@ The CLI utility supports two commands:
...
+    1. Determine a previous successful rootfs version by running ``ostree admin status`` command.
+    2. Deploy it by running ``ostree admin deploy <hash>`` command.
+    3. Reboot a device and start desired Apps.
 
-If ``aklite-offline run`` returns *100* , then an update has not been successful and rollback has happened. In this case you should trigger reboot followed by another ``aklite-offline run`` to complete the rollback.
+    Also, a user/client may decide not to rollback to the previous rootfs version in such the case.
+    Device is successfully booted on the updated rootfs and just the updated Apps failed to start.
+    Therefore, a user/client can just run the previous version of desired Apps.
 
+    To run desired Apps, a user should remove unwanted App versions from ``/var/sota/reset-apps`` and run ``aklite-apps run`` command.
```
[the remark this answers is not in the document]
Comment (mike-sul) [reply source/user-guide/offline-update/offline-update.rst:129]: I am not sure. I think it's going to create even more confusion, questions and support work. I think if it happens then it should be treated as any other general error, e.g. an update can just fail with a generic error.

I incline towards removing this text/doc at all.

File: source/user-guide/offline-update/offline-update.rst [hunk 3]
```diff
@@ -55,9 +55,16 @@ Use the ``aklite-offline`` CLI utility to perform an offline update:If a device
```
[the remark this answers is not in the document]
Comment (mike-sul) [reply source/user-guide/offline-update/offline-update.rst:55]: fixed in forced push.

**Comment by doanac** [comment]
Docs for 1ddc61f1bcceee649106778c03995b8ba8cffde5 are browsable at: https://ci.foundries.io/projects/fio-docs/builds/2009/docs/artifacts/html/index.html

File: source/user-guide/offline-update/offline-update.rst (hunk 1 above)
Comment (mike-sul) [reply source/user-guide/offline-update/offline-update.rst:56]: There is a note just after the item N3.

File: source/user-guide/offline-update/offline-update.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (mike-sul) [reply source/user-guide/offline-update/offline-update.rst:129]: It's not so easy to produce working shell script; it requires storing a state and it's use-case specific, different customer will implement it differently depending on requirements. For example, some of them may wanna implement a systemd one-shot service that runs `aklite-offline run` command if `aklite-offline install` was invoked before reboot.

[the remark this answers is not in the document]
Comment (doanac) [reply source/user-guide/offline-update/offline-update.rst:129]: maybe not fully-working/supported but we need to somehow capture the choices/options you are telling me here in the docs. maybe some pseudo code for one specific case? I'm just not sure a user could read this doc and build a working solution without a lot of help?

File: source/user-guide/offline-update/offline-update.rst [hunk 4]
```diff
@@ -78,16 +85,47 @@ The CLI utility supports two commands:
...
+- *101*: Restart of dockerd service is required , e.g. ``systemctl restart docker``. After the restart ``aklite-offline run`` must be invoked.
+
+The ``run`` command sets the following exit codes:
+
+- *0*: Update succeeded, a device is booted on the updated rootfs and running the updated Apps.
+- *90*: Update succeeded, an optional reboot can be performed to finalize the boot firmware update.
+- *99*: Update failed and a device rolled back to the previous version, no reboot is required.
+- *100*: Update failed and a device is rolling back to the previous version, reboot followed by ``aklite-offline run`` is required to complete the rollback.
+- *110*: Update failed and a device rolled back to the previous rootfs version but failed to determine Apps to run.
+- *120*: Update failed but Target to rollback to is unknown
```
Comment (vanmaegima) [inline source/user-guide/offline-update/offline-update.rst:106]: Minor suggestion to match what's on the note:
```suggestion
- *120*: Update failed to start the updated Apps but Target to rollback to is unknown.
```

File: source/user-guide/offline-update/offline-update.rst [hunk 5]
```diff
@@ -78,16 +85,47 @@ The CLI utility supports two commands:
...
+
+.. note::
+
+    If ``aklite-offline run`` command exits with ``110`` or ``120`` codes, then it indicates that
+    ``aklite-offline`` failed to determine the Target to roll back to after a failure to boot on
```
Comment (vanmaegima) [inline source/user-guide/offline-update/offline-update.rst:111]:
```suggestion
    ``aklite-offline`` failed to determine the Target to rollback to after a failure to boot on
```

File: source/user-guide/offline-update/offline-update.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (vanmaegima) [reply source/user-guide/offline-update/offline-update.rst:129]: @mike-scott had some ideas on creating a reference service for this (whole offline update process), we have an internal ticket to start with a flowchart for this.

[the remark this answers is not in the document]
Comment (mike-sul) [reply source/user-guide/offline-update/offline-update.rst:129]: >maybe some pseudo code for one specific case? 

I think the pseudo code is fine. I think two bash scripts, one for install, and another for run makes sense. Each of these scripts will be correct in terms of bash syntax, however they should not be considered as a working solution if bound together.

**Comment by doanac** [comment]
Docs for fa505ec5bba47cfb32b7f20daae4d64071b2461c are browsable at: https://ci.foundries.io/projects/fio-docs/builds/2014/docs/artifacts/html/index.html

**Comment by doanac** [comment]
Docs for 5d7ecdf346a938f835cf676808557fc490c692dd are browsable at: https://ci.foundries.io/projects/fio-docs/builds/2015/docs/artifacts/html/index.html

File: source/user-guide/offline-update/offline-update.rst (hunk 1 above)
Comment (mike-sul) [reply source/user-guide/offline-update/offline-update.rst]: I changed it, now, the reference is before.

**Comment by mike-sul** [comment]
@vanmaegima I rewrote the changes that you provided comments to, please, review it again.

File: source/user-guide/offline-update/offline-update.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (mike-sul) [reply source/user-guide/offline-update/offline-update.rst:114]: The goal of this PR is to align the doc to this changes https://github.com/foundriesio/aktualizr-lite/pull/231 which mostly is about a few additional exit codes related to rollback and bootloader update.

Other improvements seem like out of scope of the given effort.

File: source/user-guide/offline-update/offline-update.rst [hunk 6]
```diff
@@ -78,16 +82,35 @@ The CLI utility supports two commands:
...
-- *101* - Restart of dockerd service is required , e.g. ``systemctl restart docker``. After the restart ``aklite-offline run`` must be invoked.
-- *0* - OSTree/rootfs installation and Apps start was successful, update completed. Only ``aklite-offline run`` can return this code.
-
-If ``aklite-offline run`` returns *100* , then an update has not been successful and rollback has happened. In this case you should trigger reboot followed by another ``aklite-offline run`` to complete the rollback.
-
+.. _Post Install and Run Actions:
+
+Post Install and Run Actions
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+The ``install`` command sets the following exit codes (``echo $?``) that instruct which of the post install actions you should perform:
```
Comment (vanmaegima) [inline source/user-guide/offline-update/offline-update.rst:89]:
```suggestion
The ``install`` and ``run`` commands sets exit codes (``echo $?``) to instruct which of the post install actions you should perform.

The ``install`` command sets the following exit codes:
```

File: source/user-guide/offline-update/offline-update.rst [hunk 7]
```diff
@@ -78,16 +82,35 @@ The CLI utility supports two commands:
...
+
+- *0*: Device already runs the specified target, no update is needed.
+- *90*: Reboot is required to complete the previous boot firmware update. After reboot a client should repeat the update attempt from the beginning.
+- *100*: Reboot is required to complete installation. After reboot ``aklite-offline run`` must be invoked.
+- *101*: Restart of dockerd service is required , e.g. ``systemctl restart docker``. After the restart ``aklite-offline run`` must be invoked.
+
+The ``run`` command sets the following exit codes:
+
+- *0*: Update succeeded.
+    - Device is booted on the updated rootfs and running the updated Apps.
```
Comment (vanmaegima) [inline source/user-guide/offline-update/offline-update.rst:99]: The formatting of these two sections are different in the published page, can we standardize that?

File: source/user-guide/offline-update/offline-update.rst [hunk 8]
```diff
@@ -49,16 +49,20 @@ Performing the Offline Update
...
-    b. reboot device and invoke ``aklite-offline run [--config <config dir or file>]`` to finalize an ostree installation and start Apps if both ostree/rootfs and Apps are updated;
-    c. restart the Docker Engine (e.g. ``systemctl restart docker``) and invoke ``aklite-offline run [--config <config dir or file>]``  if just Apps are updated.
+    a. reboot device and invoke ``aklite-offline run [--config <config dir or file>]`` to finalize an ostree installation and start Apps if both ostree/rootfs and Apps are updated;
+    b. restart the Docker Engine (e.g. ``systemctl restart docker``) and invoke ``aklite-offline run [--config <config dir or file>]``  if just Apps are updated.
+    c. reboot device to finalize the previous boot firmware update and go to the step #1 to start the update.
 
+3. Reboot a device after running ``aklite-offline run [--config <config dir or file>]`` command if:
+
+    a. Apps failed to start after update, you must reboot a device to complete the rollback;
+    b. the update includes a boot firmware, you can optionally reboot a device to finalize the boot firmware upgrade.
```
Comment (vanmaegima) [inline source/user-guide/offline-update/offline-update.rst:65]: I tested this today and it all works fine, but I had to scroll up and down in order to find the needed steps and corresponding exit codes. What do you think of something like:
```suggestion
    a. code 100: reboot device and invoke ``aklite-offline run [--config <config dir or file>]`` to finalize an ostree installation and start Apps if both ostree/rootfs and Apps are updated;
    b. code 101: restart the Docker Engine (e.g. ``systemctl restart docker``) and invoke ``aklite-offline run [--config <config dir or file>]``  if just Apps are updated.
    c. code 90: reboot device to finalize the previous boot firmware update and go to the step #1 to start the update.

3. Reboot a device after running ``aklite-offline run [--config <config dir or file>]`` command if:

    a. code 99: Apps failed to start after update, you must reboot a device to complete the rollback;
    b. code 90: the update includes a boot firmware, you can optionally reboot a device to finalize the boot firmware upgrade.
```

File: source/user-guide/offline-update/offline-update.rst (hunk 6 above)
Comment (mike-sul) [reply source/user-guide/offline-update/offline-update.rst]: Done.

File: source/user-guide/offline-update/offline-update.rst (hunk 7 above)
Comment (mike-sul) [reply source/user-guide/offline-update/offline-update.rst]: let's how it looks now.

File: source/user-guide/offline-update/offline-update.rst (hunk 8 above)
Comment (mike-sul) [reply source/user-guide/offline-update/offline-update.rst]: done.

**Review by vanmaegima — APPROVED** [review]
LGTM

**Review by kprosise — APPROVED** [review]
LGTM, I can merge whenever you're ready!

**Comment by mike-sul** [comment]
> I can merge whenever you're ready!

It can be merged anytime, however is should not be published before LmP v90 release.

**Pull request merged**